raven_aiohttp.py

```diff
@@ -58,50 +80,129 @@ def keepalive(self):
+                                     failure_cb)
+
+    @asyncio.coroutine
+    def _do_send(self, url, data, headers, success_cb, failure_cb):
+        if self.keepalive:
+            session = self._client_session
+        else:
+            session = self._client_session_factory()
+
+        resp = None
+        try:
+            with async_timeout.timeout(self.timeout, loop=self._loop):
+                # timeout=None disables built-in aiohttp timeout
+                resp = yield from session.post(url, data=data, compress=False,
+                                               headers=headers, timeout=None)
 
-    def async_send(self, url, data, headers, success_cb, failure_cb):
-        @asyncio.coroutine
-        def f():
-            session = self.client_session()
-            try:
-                resp = yield from asyncio.wait_for(
-                    session.post(url,
-                                 data=data,
-                                 compress=False,
-                                 headers=headers),
-                    self.timeout,
-                    loop=self._loop)
-                yield from resp.release()
                 code = resp.status
                 if code != 200:
                     msg = resp.headers.get('x-sentry-error')
                     if code == 429:
                         try:
-                            retry_after = int(resp.headers.get('retry-after'))
+                            retry_after = resp.headers.get('retry-after')
+                            retry_after = int(retry_after)
                         except (ValueError, TypeError):
                             retry_after = 0
                         failure_cb(RateLimited(msg, retry_after))
                     else:
                         failure_cb(APIError(msg, code))
                 else:
                     success_cb()
-            except Exception as exc:
+        except asyncio.CancelledError:
+            # do not mute asyncio.CancelledError
+            raise
+        except Exception as exc:
+            failure_cb(exc)
+        finally:
+            if resp is not None:
+                resp.release()
+            if not self.keepalive:
+                yield from session.close()
+
+    def async_send(self, url, data, headers, success_cb, failure_cb):
+        if self._closing:
+            raise RuntimeError('AioHttpTransport is closing')
+
+        if self._background_workers:
+            data = url, data, headers, success_cb, failure_cb
+
+            try:
+                self._queue.put_nowait(data)
+            except asyncio.QueueFull as exc:
+                while True:
+                    skipped = self._queue.get_nowait()
+
+                    if skipped is ...:
+                        self._queue.put_nowait(...)
+                        continue
+
+                    break
+
+                *_, failure_cb = skipped
+
                 failure_cb(exc)
```

Comment:
Very non-obvious exception message.
Maybe we need dedicated exception type for this?

Reply:
right now it does same for any unexpected error
https://github.com/getsentry/raven-aiohttp/blob/master/raven_aiohttp.py#L99

how dedicated error can help in such case?

Reply:
What about new custom exception?

Reply:
+1